### ISSUE TITLE
Apps: Add ability to reconfigure app

### ISSUE THREAD

**Issue by asizemore** [VEuPathDB/web-eda#1109]
Each app has some set of parameters (possibly only a user-given name) that a user configures when adding the app.

Currently we don't have a way for the user to re-configure the app. Instead, once it's added, the app configuration cannot be changed.

There's a small gear icon next to the user-given app name on the thumbnail page and in the app titles. For this ticket, let's have that gear icon take a user back to the configuration page. 

Note:
- this re-configuration would need to change the verbiage and function of the (current) Add App button. Instead it should say something like "Save changes" ? Also the function on click should not add an entire new app, but instead just modify the existing app. 
- I'm assuming it's easier and clearer to send the user back to a modified version of our configuration page than opening some type of modal when the user clicks the gear icon.

**Comment by dmfalke** [comment]
I agree that we should reuse the existing configuration page. Perhaps we need to include an optional `visualizationId` prop, and use the presence of that to change the verbiage and behavior.

**Comment by asizemore** [comment]
oh that makes sense!
it also makes me realize after reconfiguring, the visualizations will have to each be opened in order to sync with the configuration changes. So vizs will need to know if they're out of sync with the current subset, or out of sync with the current configuration?
I'm not familiar with the out-of-sync-viz things - is that different enough that it should be a new issue?

**Comment by dmfalke** [comment]
That's a good question. An app instance could have many visualizations. I think we can just mark all such visualizations as out of date when the app config gets updated. I don't think we need a new issue for that. It should be done as a part of resolving this issue.

**Comment by dmfalke** [comment]
I spoke too soon. We aren't tracking out-of-sync-ness the way I thought we were. That might need to change.

**Comment by dmfalke** [comment]
The current method stores the set of filters used when the viz was last "run". We can follow this approach and also add the app config, but then we're duplicating a bit of data, and we're adding a new property to the viz. It also can have false negatives when keys change position, etc.

Another approach we can take is to generate a checksum based on the info the viz uses (current filters and app config). We would only need to store this checksum in the viz. This has the benefit of allowing us to generate a stable checksum, regardless of key ordering. It also allows us to add more dependencies without having to modify the viz object.

We should probably discuss this in our Wednesday meeting.

**Comment by dmfalke** [comment]
I have one more comment on this topic. It is unlikely we will find an encoding that will save us much space, unless we roll our own, which is not something I'm enthusiastic about. Another approach to consider is that we simply add property called `outOfDate` to the viz object, and set this to `true` when certain changes are made to the analysis object.

One downside of this approach is, if a user adds a filter, and then removes it before going back to their visualizations, they will all still be marked as out of sync.

**Comment by asizemore** [comment]
Very interesting. Another upside imo of the `outOfDate` approach is that then we don't need different 'out of sync' verbiage. If a user changes filters and the app config, we dont have to decide if the vizs should say 'out of sync with subset' or 'out of sync with app config'. It would just say a nicer version of 'stale. click to sync'. 

Looking forward to discussing tomorrow!

**Comment by dmfalke** [comment]
Yep, I agree! We'll see if others think we need more detailed verbiage.

**Comment by asizemore** [comment]
From front end meeting today: happy with the `outOfDate` approach. @danicahelb is planning to add it to this Friday's UX meeting agenda to double check that clinepi is also happy with the change

**Comment by danicahelb** [comment]
Feedback from 5/6/2022 EDA UX meeting:

Use case: user mistakenly has clicked something on the browse & subset tab and reverses the change. If the proposed simplification were to be implemented, visualizations will be flagged as out of sync as soon as the subset changes at all (even if the changes to the subset were reverted by the user)

Feedback from the group: this proposed change would be BAD! Mistakenly changing the subset and correcting your change would nuke all of your graphics

We decided that we will not move forward with the simplification of out-of-sync handling. @dmfalke will figure out the best way to handle the implementation details

**Comment by dmfalke** [comment]
I propose we use a hash of the dependencies that a visualization needs to retrieve data and render the graphics. I think either md5 or sha1 will be sufficient, in terms of avoiding collisions. We can use the library [object-hash](https://www.npmjs.com/package/object-hash), and pass in an object including the subset and the app config.

**Comment by asizemore** [comment]
Awesome. Thanks @dmfalke ! Is that still a "medium cost" approach? Or should we update the label?

**Comment by dmfalke** [comment]
I think it's still medium cost

**Comment by asizemore** [comment]
moving out of next cycle until we are more sure of our path forward with the mbio eda flow.
